src/engine/TransitivePathImpl.h

```diff
@@ -235,6 +238,201 @@ class TransitivePathImpl : public TransitivePathBase {
+    Node startNodes_;
+    TripleComponent target_;
+    bool yieldOnce_;
+    // runtime state
+    ad_utility::Timer timer_{ad_utility::Timer::Stopped};
+    LocalVocab targetHelper_;
+    LocalVocab mergedVocab_;
+    std::optional<Id> targetId_;
+    bool sameVariableOnBothSides_;
+    bool endsWithGraphVariable_;
+    bool startsWithGraphVariable_;
+    // range state
+    bool finished_{false};
+    // The `resultRange_` yields results of this transitive hull
+    // computation. Values from `TableColumnWithVocab::expandUndef` are
+    // processed one by one. The range and iterator are reinitialised when the
+    // next value from the `enumerateRange_` is yielded. The iterator is
+    // incremented on each iteration of this object.
+    std::optional<NodeWithTargetRange> resultRange_;
+    NodeWithTargetRange::iterator result_{};
+    // The `enumerateRange_` yields `pair<index, ZippedType>` for the current
+    // table column's `startNodes_`.  The range and iterator are
+    // reinitialised when the next table column is yielded. The iterator is
+    // incremented only when the `resultRange_` is exhausted.
+    std::optional<EnumerateRangeType> enumerateRange_;
+    ql::ranges::iterator_t<EnumerateRangeType> enumerateRangeIt_{};
+    // This iterator points to the table columns of `startNodes_` and only
+    // incremented when the `enumerateRange_` is exhausted. The transitive hull
+    // computation is completed when  this iterator reaches the end of
+    // `startNodes_`.
+    ql::ranges::iterator_t<Node> tableColumnIt_{};
+
+   public:
+    TransitiveHullLazyRange(const TransitivePathImpl<T>& parent, const T edges,
+                            LocalVocab edgesVocab, Node startNodes,
+                            TripleComponent start, TripleComponent target,
+                            bool yieldOnce)
+        : parent_(parent),
+          edges_(std::move(edges)),
+          edgesVocab_(std::move(edgesVocab)),
+          startNodes_(std::move(startNodes)),
+          target_(std::move(target)),
+          yieldOnce_(yieldOnce) {
+      // `targetId` is only ever used for comparisons, and never stored in the
+      // result, so we use a separate local vocabulary - targetHelper_.
+      const auto& index = parent_.getIndex();
+      targetId_ = target_.isVariable()
+                      ? std::nullopt
+                      : std::optional{std::move(target_).toValueId(
+                            index.getVocab(), targetHelper_,
+                            index.encodedIriManager())};
+      sameVariableOnBothSides_ =
+          !targetId_.has_value() && parent_.lhs_.value_ == parent_.rhs_.value_;
+      endsWithGraphVariable_ = !targetId_.has_value() &&
+                               parent_.graphVariable_ == target_.getVariable();
+      startsWithGraphVariable_ =
+          start.isVariable() && parent_.graphVariable_ == start.getVariable();
+    }
+
+    std::optional<Id> getTargetId(const Id& startNode,
```

Comment:
Please add a comment to document what this function does. Basically it handles the case where you have a transitive path like this_
```sparql
SELECT * {
  ?x <a>+ ?x . # sameVariableOnBothSides
  GRAPH ?g {
    ?y <b>+ ?g # endsWithGraphVariable
  }
  # else
  VALUES ?z { <d> }
  ?z <c>+ <e> # where <e> would be the target variable, or it would be nullopt if <e> would be ?e instead.
}
```